### ISSUE TITLE
Frequent mounts/unmounts lead to high pid1 CPU usage and reduced pid1 responsiveness

### ISSUE THREAD

**Issue by fsateler** [systemd/systemd#8703]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> 238

### Used distribution

> Debian

### In case of bug report: Expected behaviour you didn't see

> CPU usage does not rise, zombie processes do not accumulate

### In case of bug report: Unexpected behaviour you saw

> CPU usage rising significantly (anywhere between 5%-20%), zombie count rising as pid1 does not reap them in time. 

### In case of bug report: Steps to reproduce the problem

```
$ cat test.sh
#!/bin/sh

target=`mktemp -d`
sudo mount -o size=1k -t tmpfs tmpfs $target
# let's wait a bit for mountinfo to grow and be slower to parse
sleep 1
sudo umount $target
rmdir $target
$ seq 1 2000 | parallel -j1000 --progress ./test.sh
```

You may need to tweak the number of parallel jobs and total number of jobs depending on the power of your machine.

The `systemd --user` instances also respond to the same events, so CPU usage grows with the number of instances running.

A pattern of frequent mounts/unmounts can be quite common in build farms, where builds are run in namespaced containers.

**Comment by fsateler** [comment]
> So, mount event handling is notoriously awful on Linux, as we always have to rescan the full mount table. It simply doesn't scale kernel-level, and there's little we can do in userspace along to ameliorate that fully.

But it would also be possible to throttle the mountinfo parsing, to limit the CPU usage, no?

> > A pattern of frequent mounts/unmounts can be quite common in build farms, where builds are run in namespaced containers.
>
> Uh? Namespaced containers should not affect the host's mount table. Why would container mounts leak into the host file system namespace? If they do this appears like a scalability bug in the container manager in question?

Right. It was pointed out during debugging that the manager should hide the mounts. But this wouldn't work if the tool were using plain chroots.

**Comment by poettering** [comment]
> But it would also be possible to throttle the mountinfo parsing, to limit the CPU usage, no?

Yeah, I guess. Happy to review/merge a patch for that!

**Comment by poettering** [comment]
I guess this can be closed now since d586f642fd90e3bb378f7b6d3e3a64a753e51756 and the later fixes for it.
